File: backend/app/schemas/teacher.py
from datetime import datetime, date as date_type
from enum import Enum
from app.schemas.base import CamelModel


# ============================================================
#  Private Notes
# ============================================================
class NoteCreate(CamelModel):
    content: str


class NoteUpdate(CamelModel):
    content: str


class NoteResponse(CamelModel):
    id: str
    student_id: str
    teacher_id: str
    teacher_name: str | None
    content: str
    created_at: datetime
    updated_at: datetime


# ============================================================
#  Assessment Marks
# ============================================================
class MarkCreate(CamelModel):
    assessment_title: str
    score: float
    max_score: float
    comments: str | None = None
    date: date_type | None = None


class MarkResponse(CamelModel):
    id: str
    student_id: str
    project_id: str | None
    assessment_title: str
    score: float
    max_score: float
    comments: str | None
    teacher_id: str
    teacher_name: str | None
    date: date_type | None
    created_at: datetime
    updated_at: datetime


# ============================================================
#  Project Milestones
# ============================================================
class MilestoneStatus(str, Enum):
    on_track = "On Track"
    delayed = "Delayed"
    completed = "Completed"


class MilestoneCreate(CamelModel):
    description: str
    status: MilestoneStatus
    date: date_type | None = None


class MilestoneResponse(CamelModel):
    id: str
    project_id: str
    description: str
    status: str
    date: date_type | None
    created_at: datetime
    updated_at: datetime

class GuidanceCaseCreate(CamelModel):
    student_id: str
    trigger_signal: str | None = None


class GuidanceCaseUpdate(CamelModel):
    status: str | None = None
    resolution_note: str | None = None


class GuidanceCaseResponse(CamelModel):
    id: str
    student_id: str
    owning_teacher_id: str | None
    trigger_signal: str | None
    status: str
    resolution_note: str | None
    date_opened: date_type | None
    date_resolved: date_type | None
    created_at: datetime
    updated_at: datetime

File: src/api/entities/teacher.ts
/**
 * Teacher Portal specific entities.
 */

export type GuidanceCaseStatus = 'Open' | 'Assigned' | 'In Progress' | 'Resolved' | 'Closed';
export type PerformanceTier = 'High Performing' | 'Average - Guidable' | 'Underperforming';

/** 
 * Lightweight student entity for listings. 
 * Prevents fetching heavy portfolios for dashboard tables.
 */
export interface StudentSummary {
  id: string;
  prn: string;
  name: string;
  cgpa: number;
  performanceTier: PerformanceTier;
  guidanceStatus: GuidanceCaseStatus | null;
  lastInteractionDate: string | null;
  avatar?: string | null;
}

export interface PrivateNote {
  id: string;
  studentId: string;
  teacherId: string;
  teacherName: string;
  content: string;
  createdAt: string;
  updatedAt: string;
}

export interface AssessmentMark {
  id: string;
  studentId: string;
  projectId?: string | null; // null = general mark, string = project-level mark
  assessmentTitle: string;
  score: number;
  maxScore: number;
  comments: string;
  teacherId: string;
  teacherName: string;
  date: string;
  createdAt: string;
  updatedAt: string;
}

export type MilestoneStatus = 'On Track' | 'Delayed' | 'Completed';

export interface ProjectMilestone {
  id: string;
  projectId: string;
  description: string;
  status: MilestoneStatus;
  date: string;
  createdAt: string;
  updatedAt: string;
}

export interface GuidanceCase {
  id: string;
  studentId: string;
  studentName: string;
  studentPrn: string;
  studentCgpa: number;
  triggerSignal: string; // e.g., 'CGPA < 6.0', 'No interaction in 30 days'
  owningTeacherId: string;
  owningTeacherName: string;
  status: GuidanceCaseStatus;
  resolutionNote?: string;
  dateOpened: string;
  dateResolved?: string;
  createdAt: string;
  updatedAt: string;
}

export interface SupportSignal {
  id: string;
  studentId: string;
  studentName: string;
  studentPrn: string;
  studentCgpa: number;
  urgencyLabel: string; // e.g., "Critical", "Warning"
  reasonTags: string[]; // e.g., ["CGPA Declining", "No Recent Interaction"]
  isActiveSupport: boolean;
  createdAt: string;
}
